### PR TITLE
dbsp_adapters: Add input adapter for URLs.

### PR DESCRIPTION
**Pull request by blp**
I'm having a fight with the CI on this. So far (crossing fingers for latest rebase) I don't think that the failures are related to my changes.

### CODE REVIEW

**Comment by blp** [comment]
Hey, CI passed! Now looking for reviews.

File: crates/adapters/src/transport/url.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    }
+}
+
+impl Drop for UrlInputEndpoint {
+    fn drop(&mut self) {
+        self.disconnect();
+    }
+}
+
+fn rustls_config() -> Arc<ClientConfig> {
```
Comment (mihaibudiu) [inline crates/adapters/src/transport/url.rs:131]: what does this name stand for?

**Review by mihaibudiu — APPROVED** [review]

File: crates/adapters/src/transport/url.rs (hunk 1 above)
Comment (mihaibudiu) [reply crates/adapters/src/transport/url.rs:131]: shouldn't the name be rusttls?

Comment (blp) [reply crates/adapters/src/transport/url.rs:131]: > shouldn't the name be rusttls?

Feel free to file a bug upstream: https://github.com/rustls/rustls/issues

**Review by gz — APPROVED** [review]

File: crates/adapters/src/transport/url.rs (hunk 1 above)
Comment (ryzhyk) [inline crates/adapters/src/transport/url.rs:131]: @lalithsuresh , just wanted to draw your attention to this.  I am guessing we'll want to unify TLS configuration for the entire pipeline somewhere eventually.

File: crates/adapters/src/transport/url.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            // `awc` intentionally uses errors that aren't `Sync`, but
+            // `anyhow::Error` requires `Sync`.  Transform the error so we can
+            // return it.
+            |error| anyhow!("{error}"),
+        )?;
+
+        while *receiver.borrow() != PipelineState::Terminated {
+            select! {
+                _ = receiver.changed() => (),
+                result = response.next(), if *receiver.borrow() == PipelineState::Running => {
```
Comment (ryzhyk) [inline crates/adapters/src/transport/url.rs:83]: I guess if we stay in the paused state long enough, the server will eventually drop the connection and we won't be able to continue. I guess there's not a lot we can do about it without adding too much complexity.

File: crates/adapters/src/transport/url.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                    .with_root_certificates(root_store)
+                    .with_no_client_auth(),
+            )
+        };
+    }
+
+    ROOT_STORE.clone()
+}
+
+#[cfg(test)]
```
Comment (ryzhyk) [inline crates/adapters/src/transport/url.rs:155]: It would be nice to test some corner cases:
- termination/pause/unpause while downloading
- errors: invalid url, server-side disconnect.

**Review by ryzhyk — APPROVED** [review]

File: crates/adapters/src/transport/url.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+impl InputEndpoint for UrlInputEndpoint {
+    fn connect(&mut self, mut consumer: Box<dyn InputConsumer>) -> AnyResult<()> {
+        let config = self.config.clone();
+        let receiver = self.receiver.clone();
+        let _worker = spawn(move || {
+            System::new().block_on(async move {
+                let result = Self::worker_thread(config, &mut consumer, receiver).await;
+                if let Some(error) = result.err().or_else(|| consumer.eoi().err()) {
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/url.rs]: Fixed, thanks.

File: crates/adapters/src/transport/url.rs (hunk 2 above)
Comment (blp) [reply crates/adapters/src/transport/url.rs]: I pushed a change to connect on un-pause, disconnect on pause, and reconnect but skip past data already read if un-paused again.

File: crates/adapters/src/transport/url.rs (hunk 3 above)
Comment (blp) [reply crates/adapters/src/transport/url.rs:254]: I added some tests.

File: crates/adapters/src/transport/url.rs (hunk 1 above)
Comment (lalithsuresh) [reply crates/adapters/src/transport/url.rs:230]: This looks fine. We'll need to be able to configure this externally soon and so things might get tweaked a bit when we get there.

**Comment by blp** [comment]
@ryzhyk Do you want to take a look at the new version?

**Comment by ryzhyk** [comment]
I did, looks great thank you!
